## 一、第一个 Python 程序

### 输入输出

In [2]:
print('100 + 200 = ', 100 + 200)

100 + 200 =  300


In [4]:
name = input('输入姓名')

print('你输入的是:' + name)

输入姓名 古月


你输入的是:古月


In [5]:
name

'古月'

In [6]:
print('1024 * 768 =', 1024 * 768)

1024 * 768 = 786432


#### 禁止转义

In [11]:
print('\\\t\\')

\	\


In [12]:
print(r'\\\t\\')

\\\t\\


#### 换行

In [14]:
print('line1\nline2\nline3')

line1
line2
line3


In [19]:
print('''line1
line2
line3''')

line1
line2
line3


In [88]:
birth = int(input('birth: '))

birth

birth:  18


18

## 二、Python 基础

### 数据类型和变量

#### 布尔值

In [20]:
True and False

False

In [21]:
True or False

True

In [22]:
not False

True

#### 变量

- 变量可以改变类型 -> python 为动态语言：

In [25]:
a =  123
print(a)

a = 'ABC'
print(a)

a = True
print(a)

123
ABC
True


- 变量在内存中的表示：

In [31]:
a = 'ABC'

b = a

a = 'XYZ'

print(a, b)

XYZ ABC


#### 常量除法

In [34]:
10 / 3

3.3333333333333335

In [35]:
10 // 3


3

In [36]:
10 % 3

1

#### 练习

请打印出以下变量的值：

```
n = 123
f = 456.789
s1 = 'Hello, world'
s2 = 'Hello, \'Adam\''
s3 = r'Hello, "Bart"'
s4 = r'''Hello,
Lisa!'''

```

In [44]:
print('n = ', 123)
print('f = ', 456.789)
print('s1 =', 'Hello, world')
print('s2 = ', 'Hello, \\\'Adam\\\'')
print('s3 = ', r'Hello, "Bart"')
print('s4 = ', r'''Hello
Lisa''')

n =  123
f =  456.789
s1 = Hello, world
s2 =  Hello, \'Adam\'
s3 =  Hello, "Bart"
s4 =  Hello
Lisa


**注意：** 
Python的整数没有大小限制，而某些语言的整数根据其存储长度是有大小限制的，例如Java对32位整数的范围限制在-2147483648-2147483647。

Python的浮点数也没有大小限制，但是超出一定范围就直接表示为inf（无限大）。

### 字符串与编码

#### 编码介绍

|  字符   | 字节数 | 用途  |
| ----- | :------: | :-----:|
| ASCII | 1 | 英文 |
| Unicode | 2 | 全球 |
| Unicode | 4 | 生僻字 |
| UTF-8 | 可变长（1~6） | 节省空间，方便传输 |

在计算机内存中，统一使用 Unicode 编码，当需要保存到硬盘或者需要传输的时候，就转换为 UTF-8 编码。

用记事本编辑的时候，从文件读取的UTF-8字符被转换为Unicode字符到内存里，编辑完成后，保存的时候再把Unicode转换为UTF-8保存到文件：

 ![](https://img.arctee.cn/picgo/202201050408940.png)
 
 浏览网页的时候，服务器会把动态生成的Unicode内容转换为UTF-8再传输到浏览器：
 
 ![](https://img.arctee.cn/picgo/202201050409185.png)
 
 所以你看到很多网页的源码上会有类似<meta charset="UTF-8" />的信息，表示该网页正是用的UTF-8编码。

#### .py 源文件开头声明 UTF-8 编码 

由于Python源代码也是一个文本文件，所以，当你的源代码中包含中文的时候，在保存源代码时，就需要务必指定保存为UTF-8编码。

通常在文件开头写上这两行：

```
#!/usr/bin/env python3
告诉Linux/OS X系统，这是一个Python可执行程序，Windows系统会忽略这个注释；

# -*- coding: utf-8 -*-
告诉Python解释器，按照UTF-8编码读取源代码，否则，你在源代码中写的中文输出可能会有乱码。
```

申明了UTF-8编码并不意味着你的.py文件就是UTF-8编码的，必须并且要确保文本编辑器正在使用UTF-8 without BOM编码：

![](https://img.arctee.cn/picgo/202201050418214.png)


#### 格式化输出字符串

In [54]:
print('Hello, %s' % 'world')

print('%2d - %.2f' % (12, 34.5656))

Hello, world
12 - 34.57


In [55]:
r = 2.5
s = 3.14 * r *2

print(f'The area of a circle with radius {r} is {s:.2f}')

The area of a circle with radius 2.5 is 15.70


In [59]:
s1 = 72
s2 = 85

print('百分点：%.1f%%' % ((s2 / s1 - 1) * 100) )

百分点：18.1%


### 使用 list 和 tuple - 有序

#### list - 列表

In [66]:
classmate = ['aa', 'bb', 'cc']

classmate

['aa', 'bb', 'cc']

In [67]:
len(classmate)

3

In [68]:
classmate[0]

'aa'

In [69]:
classmate[-1]

'cc'

In [70]:
classmate.append('dd')

classmate

['aa', 'bb', 'cc', 'dd']

In [72]:
classmate.insert(0, '11')

classmate

['11', 'aa', 'bb', 'cc', 'dd']

In [73]:
classmate.pop()

'dd'

In [74]:
classmate.pop(0)

'11'

In [75]:
classmate

['aa', 'bb', 'cc']

In [76]:
classmate[1] = 'ff'

classmate

['aa', 'ff', 'cc']

In [79]:
s = ['python', 'java', ['asp', 'php'], 'c']

print(len(s))

s[2][1]

4


'php'

#### tuple - 元组

tuple 和 list 非常类似，但是 tuple 一旦初始化就不能修改

_增删改_ 都失效，只剩下 _查_

不可变的 tuple 有什么意义？

因为 tuple 不可变，所以代码更安全。

如果可能，能用 tuple 代替 list 就尽量用 tuple。

In [80]:
classmate2 = ('aa', 'bb', 'cc')

classmate2

('aa', 'bb', 'cc')

In [81]:
classmate2[1]

'bb'

**注意：**只有 1 个元素的 tuple 定义时必须加一个逗号,，来消除歧义：

In [82]:
t = (1, )

t

(1,)

### 条件判断 

In [87]:
age = 2

if age >= 18:
    print('your age is', age)
    print('adult')
elif age >= 6:
    print('tennager')
else:
    print('kid')

kid


### 循环

In [100]:
names = ['aa', 'bb', 'cc']

for name in names:
    print('Hello, %s!' % name)

Hello, aa!
Hello, bb!
Hello, cc!


In [90]:
range(5)

range(0, 5)

In [91]:
list(range(5))

[0, 1, 2, 3, 4]

In [96]:
sum = 0
for n in range(101):
    sum +=n

print(sum)

5050


In [98]:
sum = 0
n = 99

while n > 0:
    sum += n
    n -= 2

print(sum)

2500


### 使用 dict 和 set 

#### dict

dict 全称 dictionary，在其他语言中也称为 map，使用「键-值」（key-value）存储，具有极快的查找速度。(哈希算法)

In [103]:
d = {'aa': 89, 'bb': 56, 'cc': 99}

d['aa']

89

In [104]:
d['dd'] = 90

d

{'aa': 89, 'bb': 56, 'cc': 99, 'dd': 90}

In [105]:
'vv' in d

False

In [106]:
d.get('aa')

89

In [109]:
print(d.get('ff'))

None


In [110]:
d.get('ff', -1)  # 不存在 key 时，指定返回值。否则返回 None

-1

In [119]:
d.pop('bb')

56

In [120]:
d

{'cc': 99, 'dd': 90}

#### set

In [121]:
# 要创建一个 set，需要提供一个 list 作为输入集合：

s = set([1, 2, 3])

s

{1, 2, 3}

In [123]:
s = set([1, 2, 3, 3])

s

{1, 2, 3}

In [125]:
s.add(4)

s

{1, 2, 3, 4}

In [126]:
s.add(4)

s

{1, 2, 3, 4}

In [129]:
s.remove(1)

s

{3, 4}

In [130]:
s1 = set([1, 2, 3])
s2 = set([2, 3, 4])

s1 & s2

{2, 3}

In [131]:
s1 | s2

{1, 2, 3, 4}

#### 不可变对象

对于不变对象来说，调用对象自身的任意方法，也不会改变该对象自身的内容。相反，这些方法会创建新的对象并返回，这样，就保证了不可变对象本身永远是不可变的。

In [132]:
# 可变对象 list

a = ['c', 'a', 'b']
a.sort()

a

['a', 'b', 'c']

In [133]:
# 不可变对象 str

s = 'aBC'
s.replace('a', 'A')

s

'aBC'

In [134]:
s2 = s.replace('a', 'A')

s2

'ABC'

## 三、函数

### 调用函数

In [137]:
# 函数名其实就是指向一个函数对象的引用，完全可以把函数名赋给一个变量，相当于给这个函数起了一个“别名”：

a = abs

a(-1)

1

### 定义函数

In [138]:
def my_abs(x):
    if x >= 0:
        return x
    else:
        return -x

print(my_abs(-3))

3


In [139]:
def nop():
    pass

pass 语句什么都不做，那有什么用？实际上 pass 可以用来作为占位符，比如现在还没想好怎么写函数的代码，就可以先放一个 pass，让代码能运行起来。

#### 参数检查

In [140]:
my_abs(1, -2)

TypeError: my_abs() takes 1 positional argument but 2 were given

以上参数数量不对，python 解释器会自动检查出来 TypeError 错误。但是如果参数类型不对，解释器就无法检查出了：

In [142]:
my_abs('as')

TypeError: '>=' not supported between instances of 'str' and 'int'

In [143]:
abs('ab')

TypeError: bad operand type for abs(): 'str'

当传入了不恰当的参数时，内置函数 abs 会检查出参数错误，而我们定义的my_abs 没有参数检查，会导致 if 语句出错，出错信息和 abs 不一样。所以，这个函数定义不够完善。

In [147]:
def my_abs2(x):
    if not isinstance(x, (int, float)):
        raise TypeError('bad operand type')  # 抛出异常
    if x >= 0:
        return x
    else:
        return -x

In [146]:
my_abs2('as')

TypeError: bad operand type

In [152]:
import math 

def move(x, y, step, angle=0):
    nx = x + step * math.cos(angle)
    ny = x + step * math.sin(angle)
    return nx, ny

In [154]:
x, y = move(10, 10, 5, math.pi / 6)

print(x, y)

14.330127018922195 12.5


In [155]:
r = move(10, 10, 5 , math.pi / 6)

print(r)

(14.330127018922195, 12.5)


Python 的函数返回多值其实就是返回一个 tuple 在语法上，返回一个 tuple 可以省略括号，

In [ ]:
import math

def quadratic(a, b, c):
    # x = math.pow((b * b - 4 * a *c), 0.5)
    x = math.sqrt(b * b - 4 * a * c)
    x1 = (-b + x) / 2 / a
    x2 = (-b - x) /2/ a
    return x1, x2

### 函数的参数

#### 默认参数

In [157]:
def enroll(name, gender, age=6, city='hz'):
    print('name:', name)
    print('gender', gender)
    print('age', age)
    print('city', city)

In [158]:
enroll('aa', 'F')

name: aa
gender F
age 6
city hz


#### 可变参数 

可变参数允许你传入 0 个或任意个参数，这些可变参数在函数调用时自动组装为一个 tuple。

In [159]:
def calc(*numbers):
    sum = 0
    for n in numbers:
        sum += n
    return sum

In [160]:
calc(1, 2, 4)

7

In [162]:
calc(4)

4

Python 允许你在 list 或 tuple 前面加一个 *号，把 list 或 tuple 的元素变成可变参数传进去

In [163]:
nums = [1, 3, 4]

calc(*nums)

8

*nums 表示把 nums 这个 list 的所有元素作为可变参数传进去。这种写法相当有用，而且很常见。

#### 关键字参数

关键字参数允许你传入 0 个或任意个含参数名的参数，这些关键字参数在函数内部自动组装为一个 dict。

In [3]:
def person(name, age, **kw):
    print('name:', name, 'age:', age, 'other: ', kw)

In [4]:
person('aa', 23)

name: aa age: 23 other:  {}


In [5]:
person('aa', 23, city='hz')

name: aa age: 23 other:  {'city': 'hz'}


In [6]:
person('Adam', 30, gender='M', job='Engineer')

name: Adam age: 30 other:  {'gender': 'M', 'job': 'Engineer'}


In [8]:
extera = {'city':'Beiging', 'job':'Engineer'}

person('Jack', 24, city=extera['city'], job=extera['job'])

name: Jack age: 24 other:  {'city': 'Beiging', 'job': 'Engineer'}


In [10]:
person('Jack', 24, **extera)

name: Jack age: 24 other:  {'city': 'Beiging', 'job': 'Engineer'}


关键字参数有什么用？它可以扩展函数的功能。比如，在 person 函数里，我们保证能接收到 name 和 age 这两个参数，但是，如果调用者愿意提供更多的参数，我们也能收到。试想你正在做一个用户注册的功能，除了用户名和年龄是必填项外，其他都是可选项，利用关键字参数来定义这个函数就能满足注册的需求。

上述的关键字参数形式，用户可以输入任意的关键字。如果想要**限制输入的关键字**，就用下面介绍的命名关键字参数。

In [14]:
# 必须用 * 分割开来

def person2(name, age, *, city, job):
    print(name, age, city, job)

In [16]:
person('da', 32)

TypeError: person() missing 2 required keyword-only arguments: 'city' and 'job'

In [15]:
person2('Jack', 34, city='Beijing', job='Engineer')

Jack 34 Beijing Engineer


In [21]:
# 或者用 可变参数 分割开来

def person3(name, age, *args, city, job):
    print(name, age, city, job)
    for arg in args:
        print(arg)

In [22]:
person3('Jack', 23, city='hz', job='Stu')

Jack 23 hz Stu


In [23]:
person3('jack', 23, 'aa', 34, city='hz', job='Stu')

jack 23 hz Stu
aa
34


In [27]:
# 命名关键字参数与默认参数结合使用

def person4(name, age=25, *, city='hz', job):
    print(name, age, city, job)


In [28]:
person4('Jack', job='Stu')

Jack 25 hz Stu


#### 参数组合

在 Python 中定义函数，可以用必选参数、默认参数、可变参数、关键字参数和命名关键字参数，这5种参数都可以组合使用。但是请注意，参数定义的顺序必须是：必选参数、默认参数、可变参数、命名关键字参数和关键字参数。

In [33]:
def f1(a, b, c=0, *args, **kw):
    print('a =', a, 'b =', b, 'c =', c, 'args =', args, 'kw =', kw)

In [34]:
f1(1, 2)

a = 1 b = 2 c = 0 args = () kw = {}


In [35]:
f1(1, 2, c=3)

a = 1 b = 2 c = 3 args = () kw = {}


In [36]:
f1(1, 2, 3, 'a', 'b')

a = 1 b = 2 c = 3 args = ('a', 'b') kw = {}


In [37]:
f1(1, 2, 3, 'a', 'b', x=99)

a = 1 b = 2 c = 3 args = ('a', 'b') kw = {'x': 99}


In [38]:
args = (1, 2, 3, 4)
kw = {'d': 99, 'x': '#'}

f1(*args, **kw)

a = 1 b = 2 c = 3 args = (4,) kw = {'d': 99, 'x': '#'}


In [ ]:
def mul(x, *args):
    for arg in args:
        x *= arg
    return x

### 递归函数

In [2]:
# n!

def fact(n):
    if n == 1:
        return 1
    return n * fact(n-1)

In [3]:
fact(3)

6

In [4]:
fact(5)

120

In [5]:
fact(100)

93326215443944152681699238856266700490715968264381621468592963895217599993229915608941463976156518286253697920827223758251185210916864000000000000000000000000

如果我们计算 fact(5)，可以根据函数定义看到计算过程如下：
```
===> fact(5)
===> 5 * fact(4)
===> 5 * (4 * fact(3))
===> 5 * (4 * (3 * fact(2)))
===> 5 * (4 * (3 * (2 * fact(1))))
===> 5 * (4 * (3 * (2 * 1)))
===> 5 * (4 * (3 * 2))
===> 5 * (4 * 6)
===> 5 * 24
===> 120
```

由上可见，每 return 一次，就会产生一个栈帧。而且每个栈帧必须在内存中保留到最后一刻，因为直到 120 计算出来后，fact(5) 的值还是不知道的，还需要逐层返回，逐层消栈帧。这也就可能导致栈溢出。

In [ ]:
# 使用递归，若层数过大会导致栈溢出

#fact(10000)

解决递归调用栈溢出的方法是通过尾递归优化，事实上尾递归和循环的效果是一样的，所以，把循环看成是一种特殊的尾递归函数也是可以的。

尾递归是指，在函数返回的时候，调用自身，并且，**return 语句不能包含表达式。**

这样，编译器或者解释器就可以把尾递归做优化，使递归本身无论调用多少次，都只占用一个栈帧，不会出现栈溢出的情况。

上面的 fact(n) 函数由于 return n * fact(n - 1) 引入了乘法表达式，所以就不是尾递归了。

要改成尾递归方式，需要多一点代码，主要是要把每一步的乘积传入到递归函数中：

In [3]:
def fact(n):
    return fact_iter(n, 1)

def fact_iter(num, product):
    if num == 1:
        return product
    return fact_iter(num-1, num * product)

In [5]:
fact(5)

120

可以看到，return fact_iter(num - 1, num * product) 仅返回递归函数本身，num - 1 和 num * product 在函数调用前就会被计算，不影响函数调用。

fact(5) 对应的 fact_iter(5, 1) 的调用如下：
```
===> fact_iter(5, 1)
===> fact_iter(4, 5)
===> fact_iter(3, 20)
===> fact_iter(2, 60)
===> fact_iter(1, 120)
===> 120
```

由上可见，每次的 return 已经包含了需要的历史值，如 60，并不需要再给上层返回值。前面的 return 其实可以忽略掉，只保留最后一条 return 就行。（我觉得这就是优化的意思吧）

尾递归调用时，如果做了优化，栈不会增长，因此，无论多少次调用也不会导致栈溢出。

遗憾的是，大多数编程语言没有针对尾递归做优化，Python 解释器也没有做优化，所以，即使把上面的 fact(n) 函数改成尾递归方式，也会导致栈溢出。

In [24]:
# 汉诺塔

def move(n, fro, buffer, to):
    if n == 1:
        print('from', fro, '--->', to) # 只有一个，直接从起点移到终点
    else:
        move(n-1, fro, to, buffer) # 上面的 2~n 个先移到缓冲区
        move(1, fro , buffer, to) # 最下面的一个直接移到终点
        move(n-1, buffer, fro, to) # 再把 2~n 个从缓冲区移到终点

In [25]:
move(3, 'A', 'B', 'C')

from A ---> C
from A ---> B
from C ---> B
from A ---> C
from B ---> A
from B ---> C
from A ---> C


In [21]:
# c 语言

void move(int n, char a, char b, char c)
{
    if (n == 1) {
        printf("from %c to %c\n", a, c);
    }
    else {
        move(n - 1, a, c, b);
        move(1, a, b, c);
        move(n-1, b, a, c);
    }
}

SyntaxError: invalid syntax (Temp/ipykernel_12792/4131952417.py, line 3)

In [19]:
move(3, 'A', 'B', 'C')

Move 1 from A to C
Move 1 from A to B
Move 1 from C to B
Move 1 from A to C
Move 1 from B to A
Move 1 from B to C
Move 1 from A to C


## 四、高级特性

以上学习了 Python 的数据类型、判断和循环语句、函数，已经可以写出很多有用的程序了。

只是此时的代码可能不精简、同样的问题会写很多行代码来解决。但是，Python 的思想是「代码越少越好，越简单越好」，那么就要学习 Python 的高级特性来帮助我们精简自己的代码，来用更少的代码解决问题。

学习高级特性之前，要先搞明白为什么学它。

接下来要学习的高级特性：

- 切片
- 迭代
- 列表生成式
- 生成器
- 迭代器

**在 Python 中，代码不是越多越好，而是越少越好。**代码不是越复杂越好，而是越简单越好。

基于这一思想，来学习 Python 中非常有用的**高级特性**，1 行代码能实现的功能，决不写 5 行代码。**请始终牢记，代码越少，开发效率越高。**

In [32]:
# 代码多

L = []

n = 1
while n <= 9:
    L.append(n)
    n += 2
    
L

[1, 3, 5, 7, 9]

In [35]:
# 精简

L1 = list(range(1, 9, 2))

L1

[1, 3, 5, 7]

### 切片
- 便于取子部份

In [36]:
L = ['aa', 'bb', 'cc', 'dd']


In [37]:
# 取前 n 个元素

# 原来实现方法：

r = []
n = 3

for i in range(n):
    r.append(L[i])

r

['aa', 'bb', 'cc']

In [38]:
# 简单实现

L[0:3]

['aa', 'bb', 'cc']

In [39]:
L[:3]

['aa', 'bb', 'cc']

In [1]:
L = list(range(100))

In [2]:
L[::5]

[0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95]

In [5]:
# tuple 可以切片

(0, 1, 2, 3, 4, 5)[:3]

(0, 1, 2)

In [6]:
# 字符串可以切片

'ABCDEFG'[:3]

'ABC'

In [53]:
def trim(s):
    start = 0
    end = 0
    
    # 开头空格数
    for c in s:
        if c == ' ':
            # start++ 错误，python 没有自增、自减语法
            start += 1
        else:
            break
    # 结尾空格数
    for c in s[::-1]:
        if c == ' ':
            end += 1
        else:
            break
    return s[start : len(s)-end]

In [60]:
# 测试:
if trim('hello  ') != 'hello':
    print('测试失败!')
elif trim('  hello') != 'hello':
    print('测试失败!')
elif trim('  hello  ') != 'hello':
    print('测试失败!')
elif trim('  hello  world  ') != 'hello  world':
    print('测试失败!')
elif trim('') != '':
    print('测试失败!')
elif trim('    ') != '':
    print('测试失败!')
else:
    print('测试成功!')

测试成功!


In [64]:
# 更好的实现方法

def trim2(s):
    
    while s[:1] == ' ':  # 第一个元素 s[0] 是空格
        s = s[1:]  # 舍弃第一个元素，从第二个元素 s[1] 开始
    while s[-1:] == ' ': # 最后一个元素 s[-1] 是空格
        s = s[0:-1]  # 舍弃最后一个元素
    return s

In [65]:
# 测试:
if trim2('hello  ') != 'hello':
    print('测试失败!')
elif trim2('  hello') != 'hello':
    print('测试失败!')
elif trim2('  hello  ') != 'hello':
    print('测试失败!')
elif trim2('  hello  world  ') != 'hello  world':
    print('测试失败!')
elif trim2('') != '':
    print('测试失败!')
elif trim2('    ') != '':
    print('测试失败!')
else:
    print('测试成功!')

测试成功!


### 迭代(Iteraton)
- 便于遍历循环

In [1]:
d = {'a': 1, 'b': 2, 'c': 3}

for key in d:
    print(key)

a
b
c


In [3]:
for value in d.values():
    print(value)

1
2
3


In [8]:
for k, v in d.items():
    print(k, ':', v)

a : 1
b : 2
c : 3


如何判断一个对象是可迭代对象呢？方法是通过 collections.abc 模块的Iterable 类型判断：

In [9]:
from collections.abc import Iterable

isinstance('abc', Iterable)

True

如何对 list 同时实现索引和元素本身的迭代呢？ 
需要借助 enumerate() 函数把 list 变为 索引-元素对

In [10]:
for i, v in enumerate(['A', 'B', 'C', 'D']):
    print(i, v)

0 A
1 B
2 C
3 D


In [15]:
def findMinAndMax(L):
    if L == []:
        return (None, None)
    min = max = L[0]
    for n in L:
            if n < min:
                min = n
            if n > max:
                max = n
    return (min, max)

In [16]:
findMinAndMax([1, 3, 5, 3, 6, 45, 6, 8])

(1, 45)

In [17]:
# 测试
if findMinAndMax([]) != (None, None):
    print('测试失败!')
elif findMinAndMax([7]) != (7, 7):
    print('测试失败!')
elif findMinAndMax([7, 1]) != (1, 7):
    print('测试失败!')
elif findMinAndMax([7, 1, 3, 9, 5]) != (1, 9):
    print('测试失败!')
else:
    print('测试成功!')

测试成功!


### 列表生成式（List Comprehensions）
- 便于生成列表

若想要生成这样的 list [1x1, 2x2, 3x3, ..., 10x10]，传统做法：

In [19]:
L = []
for i in range(1, 11):
    L.append(i * i)
L

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

使用列表生成式，可以简化为一行代码：

In [21]:
[ i * i for i in range(1, 11)]

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

for循环后面还可以加上if判断，这样我们就可以筛选出仅偶数的平方：

In [22]:
[i * i for i in range(1, 11) if i % 2 == 0]

[4, 16, 36, 64, 100]

还可以使用两层循环，生成全排列：

In [23]:
[m + n for m in 'ABC' for n in 'XYZ']

['AX', 'AY', 'AZ', 'BX', 'BY', 'BZ', 'CX', 'CY', 'CZ']

In [24]:
import os

[d for d in os.listdir('.')]

['.ipynb_checkpoints',
 'python-notes.ipynb',
 'untitled.tex',
 '廖雪峰教程-代码笔记.ipynb']

In [27]:
d = {'X': 'A', 'Y': 'B', 'Z': 'C'}

[k + '=' + v for k, v in d.items()]

['X=A', 'Y=B', 'Z=C']

In [29]:
L = ['Hello', 'World', 'IBM', 'Apple']

[s.lower() for s in L]

['hello', 'world', 'ibm', 'apple']

if ... else 注意事项：
- if 在 for 后面：if 是筛选条件，不能加 else
- if 在 for 前面：if 是表达式，必须加 else

In [33]:
# if 在 for 后：

[x for x in range(1, 11) if x % 2 == 0]

[2, 4, 6, 8, 10]

In [35]:
# if 在 for 前：

[x if x % 2 == 0 else -x for x in range(1, 11)]

[-1, 2, -3, 4, -5, 6, -7, 8, -9, 10]

In [38]:
L1 = ['Hello', 'World', 18, 'Apple', None]

L2 = [s.lower() for s in L1 if isinstance(s, str)]

L2

['hello', 'world', 'apple']

### 生成器（generator）

列表生成式会一次性创建出所有元素，占用很大的内存。

而生成器是采用某种算法，一遍循环一边计算出需要的元素。这样就不必创建完整的 list，从而节省大量的空间。

#### 方法一：直接将 [] 改为 ()

In [40]:
g = (x * x for x in range(10))

g

<generator object <genexpr> at 0x00000267E5760430>

In [41]:
next(g)

0

In [42]:
next(g)

1

In [43]:
for i in g:
    print(i)

4
9
16
25
36
49
64
81


##### 方法二：如果算法比较复杂，用生成式的 for 循环无法实现时，还可以用函数来生成。

In [45]:
# 传统的斐波拉写数列

def fib(max):
    n, a, b = 0, 0, 1
    while n < max:
        print(b)
        a, b = b, a + b
        n += 1
    return 'done'

In [46]:
fib(5)

1
1
2
3
5


'done'

可以看出，fib() 函数实际上是定义了斐波拉契数列的推算规则，可以从第一个元素开始，推算出后续任意的元素，这种逻辑其实非常类似 generator。

上面的函数和 generator 仅一步之遥。要把 fib 函数变成 generator 函数，只需要把 print(b) 改为 yield b 就可以了：

In [53]:
def fib2(max):
    n, a, b = 0, 0, 1
    while n < max:
        yield b
        a, b = b, a + b
        n += 1
    return 'done'
    

In [54]:
f = fib2(6)

f

<generator object fib2 at 0x00000267E576B190>

In [51]:
for n in f:
    print(n)

1
1
2
3
5
8


In [56]:
g = fib(6)

while True:
    try:
        x = next(g)
        print('g:', x)
    except StopIteration as e:
        print('Generator return value:', e.value)
        break

g: 1
g: 1
g: 2
g: 3
g: 5
g: 8
Generator return value: done


#### 杨辉三角课后练习

In [57]:
# 暂留

### 迭代器
- 可以被 next() 函数调用并不断返回下一个值的对象称为迭代器：Iterator。

In [58]:
# 可以使用 isinstance()判断一个对象是否是Iterator对象：

from collections.abc import Iterator

isinstance([], Iterator)

False

In [59]:
isinstance((x for x in range(10)), Iterator)

True

生成器都是 Iterator 对象，但 list、dict、str 虽然是 Iterable，却不是Iterator。

把 list、dict、str 等 Iterable 变成 Iterator 可以使用 iter()函数：

In [60]:
isinstance(iter([]), Iterator)

True

小结

凡是可作用于 for 循环的对象都是 Iterable 类型；

凡是可作用于 next() 函数的对象都是 Iterator 类型，它们表示一个惰性计算的序列；

集合数据类型如 list、dict、str 等是 Iterable 但不是 Iterator，不过可以通过 iter() 函数获得一个 Iterator对象。

Python 的 for 循环本质上就是通过不断调用 next() 函数实现的，例如：

In [61]:
for x in [1, 2, 3, 4, 5]:
    pass

In [62]:
# 实际上完全等价于：

# 首先获得Iterator对象:
it = iter([1, 2, 3, 4, 5])
# 循环:
while True:
    try:
        # 获得下一个值:
        x = next(it)
    except StopIteration:
        # 遇到StopIteration就退出循环
        break

## 函数式编程

函数式编程（请注意多了一个「式」字）—— Functional Programming，虽然也可以归结到面向过程的程序设计，但其思想更接近数学计算。

纯粹的函数式编程语言编写的函数没有变量，因此，任意一个函数，只要输入是确定的，输出就是确定的，这种纯函数我们称之为没有副作用。

函数式编程的一个特点就是，允许把函数本身作为参数传入另一个函数，还允许返回一个函数！